JWT VC Issuer Metadata /.well-known/jwt-vc-issuer to comply with SD-JWT VC Specification

francis-pouatcha opened the pull request:
closes #29634

francis-pouatcha committed:
Exposing JWT VC Issuer Metadata /.well-known/jwt-vc-issuer to comply with SD-JWT VC Specification

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

wistefan reviewed (approved):
LGTM

ahus1 commented:
@mposolda - I assume this is in the area of core-clients to approve and review.

francis-pouatcha committed:
Merge branch 'keycloak:main' into issue-29634

bengo237 committed:
Merge branch 'main' into issue-29634

mposolda commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/JWTVCIssuerWellKnownProvider.java:
```diff
@@ -0,0 +1,93 @@
+        Aware of duplicate logic from; org.keycloak.protocol.oidc.OIDCLoginProtocolService.certs
+         But still to check if all oidc key also qualify as jwt-vc issuer keys.
+        */
+        JWK[] keys = session.keys().getKeysStream(realm)
```
Minor: Is it possible to introduce some new class (possibly in package `org.keycloak.protocol.oidc.utils` in keycloak-services) and add the new method like:
```
JSONWebKeySet getRealmJwks(RealmModel realm)
```
and make sure that both this one and `OIDCLoginProtocolService.certs()` call that one? Would be good to avoid code duplications.

It may be also possible to rename class `JWKSHttpUtils` to more generic `JWKSServerUtils` and add the new method there.

francis-pouatcha replied:
Fixed with: #4b1bfd7

mposolda reviewed (commented):
@francis-pouatcha @wistefan Thanks for adding this and for the review!

Added comment inline regarding to avoid code duplication.

Is it also possible to add automated tests for this?

francis-pouatcha committed:
Merge branch 'main' into issue-29634

francis-pouatcha committed:
Refactor: Extract common key transformation logic to reduce duplication.

Addresses review comment in PR #29635: https://github.com/keycloak/keycloak/pull/29635/#pullrequestreview-2070540725

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
@francis-pouatcha @wistefan Added automated test covering the new functionality and verifying that the correct issuer and JWK set are returned.

This addresses the request for tests in the review discussion: https://github.com/keycloak/keycloak/pull/29635/#pullrequestreview-2070540725

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha commented:
@mposolda not sure how to proceed here. Will wait for your review comments.

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/JWTVCIssuerWellKnownProviderTest.java:
```diff
@@ -0,0 +1,68 @@
+     @Test
+    public void getConfig() {
+        String expectedIssuer = suiteContext.getAuthServerInfo().getContextRoot().toString() + "/auth/realms/" + TEST_REALM_NAME;
+        testingClient
```
Instead of using `testingClient` here (which is probably the reason of failing test btw), can this be tested by actually sending the request to the endpoint and check that data are really available under specified URL?

For the inspiration, you can take a look at how this is tested for OIDC endpoint. There are methods `AbstractWellKnownProviderTest.getOIDCDiscoveryRepresentation` and `AbstractWellKnownProviderTest.getOIDCDiscoveryConfiguration`, which are called for instance from `OIDCWellKnownProviderTest.testDefaultProviderCustomizations()`

francis-pouatcha replied:
Thanks for the hints. Fixed with [4e6c7ec](https://github.com/keycloak/keycloak/pull/29635/commits/4e6c7ec61f01d5ec18180a7fde302d27ca554c30).

mposolda reviewed (commented):
@francis-pouatcha Thanks for the updates!

The test is actually failing. It probably works with Keycloak on embedded undertow (so I guess it works for you when you run the test from your IDE), but it fails with quarkus distribution, which can be simulated probably by command like this:
```
mvn clean install
cd testsuite/integration-arquillian
mvn clean install -Pauth-server-quarkus -Dtest=JWTVCIssuerWellKnownProviderTest
```

I believe it will help to slightly update the test and test by actually sending real HTTP request instead of using "testing client" to directly call Java API. That's how OIDC well-known endpoint is also tested. Added inline comment with some details.

francis-pouatcha committed:
Fix: Failing tests - Accessing Endpoint thru http

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Merge branch 'keycloak:main' into issue-29634

ahus1 commented:
@mposolda - tests are now green.

mposolda reviewed (approved):
@francis-pouatcha @wistefan @ahus1 Thanks everyone for the work on this PR and reviews
